### PR TITLE
SAK-50745 Site navigation improve the fidelity of archive/merge

### PR DESCRIPTION
**Pull request by csev**
This will compile but not work properly or be testable until the JIRA known as 50720 is merged.

### CODE REVIEW

File: kernel/kernel-impl/src/main/java/org/sakaiproject/site/impl/BaseSiteService.java [hunk 1]
```diff
@@ -3575,9 +3578,61 @@ public String merge(String siteId, Element el, String creatorId)
...
+		// added twice.
+		try
+		{
+			Site site = getSite(siteId);
+			boolean changed = false;
+			List<SitePage> pages = site.getOrderedPages();
+			Set<String> pageNames = new LinkedHashSet();
+			for (SitePage page : pages) {
+				pageNames.add(page.getTitle());
+			}
```
Comment (adrianfish) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/site/impl/BaseSiteService.java:3596]:
```suggestion
			Set<String> pageNames = pages.stream().map(SitePage::getTitle).collect(Collectors.toCollection(LinkedHashSet::new));
```

**Review by adrianfish — CHANGES_REQUESTED** [review]
My usual pedantic review. I like the idea of adding the pages. It was driving me mad seeing all those weird log statements about missing tools.

File: kernel/kernel-impl/src/main/java/org/sakaiproject/site/impl/BaseSiteService.java [hunk 2]
```diff
@@ -3575,9 +3578,61 @@ public String merge(String siteId, Element el, String creatorId)
...
+
+			NodeList pageNodes = el.getElementsByTagName("page");
+			for (int i=0; i<pageNodes.getLength(); ++i) {
+				Element pageEl = (Element) pageNodes.item(i);
+				String pageTitle = pageEl.getAttribute("title");
+				if ( pageTitle == null ) continue;
+				if ( pageNames.contains(pageTitle) ) continue;
```
[the remark this answers is not in the document]
Comment (csev) [reply kernel/kernel-impl/src/main/java/org/sakaiproject/site/impl/BaseSiteService.java:3603]: When I write code, absent of the "when in Rome" rule, I put more not less spaces lots of places.   I like after the last ( on an if or right before the last ( on an if.  For a brace on a try I would add a space before the {.  I would follow the rules in nearby code (including brace on same line or new line).

File: kernel/kernel-impl/src/main/java/org/sakaiproject/site/impl/BaseSiteService.java [hunk 3]
```diff
@@ -3575,9 +3578,61 @@ public String merge(String siteId, Element el, String creatorId)
...
+
+				// addPage
+				BaseSitePage page = new BaseSitePage(this, pageEl, site);
```
[the remark this answers is not in the document]
Comment (csev) [reply kernel/kernel-impl/src/main/java/org/sakaiproject/site/impl/BaseSiteService.java:3606]: I tend to do these things so I can debug them with a print statement or debugger.  This particular code was a little tricky and it was nice to verify what was coming back from the constructor.  I don't like making code too tight and trust optimizer to do the more efficient thing.

File: kernel/kernel-impl/src/main/java/org/sakaiproject/site/impl/BaseSiteService.java (hunk 1 above)
Comment (csev) [reply kernel/kernel-impl/src/main/java/org/sakaiproject/site/impl/BaseSiteService.java:3596]: I am glad you changed this after I debugged it.

Comment (adrianfish) [reply kernel/kernel-impl/src/main/java/org/sakaiproject/site/impl/BaseSiteService.java]: Lol

File: kernel/kernel-impl/src/main/java/org/sakaiproject/site/impl/BaseSiteService.java [hunk 4]
```diff
@@ -3575,9 +3578,58 @@ public String merge(String siteId, Element el, String creatorId)
...
+				site.getPages().add(page);
+				changed = true;
+			}
+
+			if ( changed ) {
+				site.regenerateIds();
+				save(site);
+				// Invalidate the user-site cache.
```
Comment (csev) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/site/impl/BaseSiteService.java:3611]: Move comment to the right place

**Pull request merged**